Feature: Support Webpack 4

asbjornh opened the pull request:
So this is my first PR and there's a high chance that I'm doing something weird in terms of forks, branches, travis etc etc. If so, bear with me and let me know 🙌

Anyhoo, I got it working with webpack 4 by replacing the `q-io` file reading with a fs implementation. 🎉 Also I had to do 'addDepencency' on loaded scss files in order for hot reloading to work.

I'm hoping that your Travis is set up to build on PRs. Tests are passing locally but my own Travis account didn't find my fork so I'm not able to verify 100%.

Fixes #2  (At least for webpack 4)

Commit:
Replace gfs with fs implementation. Add dependencies to scss files for hot reloading

Commit:
Add dependencies when files are actually found

Commit:
Fix error logging

Commit:
Add test for read-file.js

Commit:
Remove q-io dependency

Commit:
Bump version and add 'files' in package.json

Commit:
Add q-io to devDependencies again 🙃

asbjornh commented:
Ok so the test for node 0.12 is failing, but I'm getting the same error on `master`. Looks like an incompatible transitive dependency of ava rather than a failing test case

NickHeiner commented:
> So this is my first PR and there's a high chance that I'm doing something weird in terms of forks, branches, travis etc etc. 

Nah, this basically looks good to me from a mechanical perspective. Welcome!

> Anyhoo, I got it working with webpack 4 by replacing the q-io file reading with a fs implementation. 

Why was this necessary?

> Also I had to do 'addDepencency' on loaded scss files in order for hot reloading to work.

Cool!

> Ok so the test for node 0.12 is failing,

I'm fine just disabling the tests for an ancient version of Node.js. :smile: Want to update the `.travis.yml` to use more modern versions of Node.js? (There may be a spot in the README that's good to update too.)

NickHeiner commented on index.js:
```diff
@@ -2,46 +2,48 @@
+    readFile = require('./lib/read-file');
 
-module.exports = function(content) {
+module.exports = function (content) {
```
I'd prefer to leave the formatting as-is.

NickHeiner commented on lib/read-file.js:
```diff
@@ -0,0 +1,18 @@
+
+function readFile(filePath) {
+  return q.Promise(function (resolve, reject) {
+    fs.readFile(filePath, 'utf-8', function (error, fileContent) {
```
Assuming that we do need to get ride of `q-io`, let's use something like `q.denodeify` or `q.nfbind` instead of wrapping it manually.

asbjornh replied:
Ooh I wasn't aware of those. That's definitely much cleaner!

NickHeiner commented on package.json:
```diff
@@ -1,8 +1,14 @@
 {
   "name": "sassjs-loader",
-  "version": "1.0.0",
+  "version": "1.1.0",
```
If the we're now supporting Webpack 4, then presumably it no longer works with Webpack 1. So this should be a major bump, because it's breaking backwards compat.

NickHeiner replied:
Also, you should change the `webpack` version in `devDependencies` to be 4, so we're testing with the version that we want to support.

asbjornh commented:
Thanks for getting back quickly!

I'll look into your comments after dinner

> Why was this necessary?

To tell you the truth, I have no idea why `q-io` was making webpack 4 fail. I got a cryptic error related to an internal webpack optimization thing. I just tried commenting things out until I got it running without errors 😛

Here is the error:

```
/Users/asbjornh/Documents/KODE/test/node_modules/sass.js/dist/sass.sync.js:61
var Module=typeof Module!=="undefined"?Module:{};var moduleOverrides={};var key;for(key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}Module["arguments"]=[];Module["thisProgram"]="./this.program";Module["quit"]=(function(status,toThrow){throw toThrow});Module["preRun"]=[];Module["postRun"]=[];var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;if(Module["ENVIRONMENT"]){if(Module["ENVIRONMENT"]==="WEB"){ENVIRONMENT_IS_WEB=true}else if(Module["ENVIRONMENT"]==="WORKER"){ENVIRONMENT_IS_WORKER=true}else if(Module["ENVIRONMENT"]==="NODE"){ENVIRONMENT_IS_NODE=true}else if(Module["ENVIRONMENT"]==="SHELL"){ENVIRONMENT_IS_SHELL=true}else{throw new Error("Module['ENVIRONMENT'] value is not valid. must be one of: WEB|WORKER|NODE|SHELL.")}}else{ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";EN

TypeError: availableModules is not iterable
    at handler (/Users/asbjornh/Documents/KODE/test/node_modules/webpack/lib/optimize/RemoveParentModulesPlugin.js:101:23)
    at SyncBailHook.eval [as call] (eval at create (/Users/asbjornh/Documents/KODE/test/node_modules/tapable/lib/HookCodeFactory.js:17:12), <anonymous>:13:16)
    at SyncBailHook.lazyCompileHook [as _call] (/Users/asbjornh/Documents/KODE/test/node_modules/tapable/lib/Hook.js:35:21)
    at Compilation.seal (/Users/asbjornh/Documents/KODE/test/node_modules/webpack/lib/Compilation.js:897:35)
    at hooks.make.callAsync.err (/Users/asbjornh/Documents/KODE/test/node_modules/webpack/lib/Compiler.js:494:17)
    at _done (eval at create (/Users/asbjornh/Documents/KODE/test/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:9:1)
    at _err1 (eval at create (/Users/asbjornh/Documents/KODE/test/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:32:22)
    at _addModuleChain (/Users/asbjornh/Documents/KODE/test/node_modules/webpack/lib/Compilation.js:771:12)
    at processModuleDependencies.err (/Users/asbjornh/Documents/KODE/test/node_modules/webpack/lib/Compilation.js:710:9)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
error Command failed with exit code 7.
```

It doesn't even look related to `q-io`, but like I said, replacing the file reading with `fs` the error magically disappeared 🙂

Commit:
Update node testing versions

Commit:
Upgrade webpack dev dependency to v4

Commit:
Fix broken test due to use of q-io and 'mode' missing from webpack config

Commit:
Remove manual promise wrapping of fs.readFile

Commit:
Bump version to 2.0.0 and add lock file

Commit:
Remove support for node version 4 in tests (webpack 4 requires v6)

asbjornh commented:
Pushed a new update.

I didn't notice this last time because I'd forgotten to upgrade the webpack dependency, but webpack 4 isn't compatible with node 4, so i upped the test target to 6. I also think testing in `lts` is a good idea so I added that one as well.

I had to remove `q-io` from `assert-css-equal` as it produced the same error as in `index.js` when running the tests with webpack 4.

I also added a lockfile, which will hopefully ensure that tests will pass in the distant future.

asbjornh commented:
> (There may be a spot in the README that's good to update too.)

The readme only references `.travis.yml` so I think it looks good the way it is 🙂

asbjornh commented:
Also, I tried upgrading to the latest version of `sass.js`, which made the `Import from node_modules` test fail because the output is different (which I guess is to be expected).

I think it would be nice to upgrade to the latest version, but I'm not sure what to do about `node-modules-import.css.txt` which has to be updated to the new version somehow

NickHeiner committed:
Update .travis.yml

NickHeiner commented on index.js:
```diff
@@ -53,22 +55,22 @@ module.exports = function(content) {
             callback(null, result.text);
         } else {
-            callback(result);
+            callback(new Error(result.formatted));
```
👍 Thanks for making this change.

NickHeiner committed:
Update sass.js version

NickHeiner commented:
> To tell you the truth, I have no idea why q-io was making webpack 4 fail.

Wow. I pulled down the code locally and also poked around, and I have no idea either. I was amazed that the simple presence of `require('q-io/fs')` causes the problem, even if `q-io` is not actually used. So, fuck it, we can just roll with the `fs`-based replacement.

Good idea to update the sass.js version. Updating the expectations test file was easy enough; I just added a `fs.writeFileSync('out.css', actualGeneratedContent)` and then copied that in to place. I did a spot-check, and the new content seems reasonable, but who knows for sure? 🙃 (This is the problem with the type of test I put in where it's a massive, black-box assertion. :joy:)

This PR will be good to go once you remove the extraneous formatting changes.

NickHeiner commented:
Also, hopefully you aren't bothered by me pushing a few commits directly to your branch. It seemed like the fastest way to get this PR merged, which seems like something you'd be in favor of. :smile:

asbjornh commented:
Great! Thanks :)

I have no objections to you committing to my fork, but do I need to review my access settings? 😛
Or do you specifically have write privileges to anything that's forked from one of your repos? Like I said, new to this whole workflow of between-repo-PRs.

I'll take a look at those formatting changes. I'll be sure to disable Prettier next time I'm contributing to something

Commit:
Clean up whitespace in callbacks

Commit:
Clean up whitespace in object literals

Commit:
Cleans up object literal whitespace in assert-css-equal

asbjornh commented:
Alright so I've undone most of the formatting done by Prettier. I left the changes to indentation [here](https://github.com/NickHeiner/sassjs-loader/pull/3/files?utf8=%E2%9C%93&diff=unified#diff-168726dbe96b3ce427e7fedce31bb0bcL20) because it seems more correct to me. Also I kept the changes that removed unnecessary whitespace at the end of lines and on empty lines.

Happy? 🙂

NickHeiner commented:
> Or do you specifically have write privileges to anything that's forked from one of your repos?

Yes, I'm almost entirely certain that's correct. There may be a setting about this when you're submitting the PR.

NickHeiner commented:
Merged and published as 2.0.0. Thanks for your contribution!

asbjornh commented:
Yay! Thanks 😃🍻
